### PR TITLE
Object_group

### PR DESCRIPTION
**Pull request by saichint**
This PR is for adding object_group and object_group_entry providers. 
All tests pass.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2866,6 +2874,126 @@ keyword 'default'.
...
+| Example Parameter Usage
+|:--
+| `cisco_object_group { '<afi> <type> <grp_name>':`
+| `cisco_object_group { 'ipv4 address myog_v4_addr':`
+
+##### `afi`
+Address Family Identifier (AFI). Required. Valid values are ipv4 and ipv6.
+
+##### `type`
+Type of the object_group instance. Required. Valid values are address and port.
```
Comment (mikewiebe) [inline README.md]: 'address' and 'port'

File: README.md [hunk 2]
```diff
@@ -2866,6 +2874,126 @@ keyword 'default'.
 Virtual ip configuration. Valid values are an array of Strings or 'default'.
 
 --
+### Type: cisco_object_group
```
Comment (mikewiebe) [inline README.md]: So.. silly question.  Do we even need a `cisco_object_group` type?  What is the point of having one without an entry under the object group?

File: README.md [hunk 3]
```diff
@@ -2866,6 +2874,126 @@ keyword 'default'.
...
+
+cisco_object_group_entry { 'ipv6 address myoge_v6_addr 30':
+  ensure              => 'present',
+  address             => '2000::1/64',
+}
+```
+
+#### Parameters
+
+| Example Parameter Usage
```
Comment (mikewiebe) [inline README.md]: You already have the Example usage above.  Not sure this Example Parameter Usage section adds any value to that.

File: README.md [hunk 4]
```diff
@@ -2866,6 +2874,126 @@ keyword 'default'.
...
+|:--
+| `cisco_object_group_entry { '<afi> <type> <grp_name> <seqno>':`
+| `cisco_object_group_entry { 'ipv4 address myoge_v4_addr 10':`
+
+##### `afi`
+Address Family Identifier (AFI). Required. Valid values are ipv4 and ipv6.
```
Comment (mikewiebe) [inline README.md]: 'ipv4' and 'ipv6'

File: README.md [hunk 5]
```diff
@@ -2866,6 +2874,126 @@ keyword 'default'.
...
+
+##### `type`
+Type of the object_group instance. Required. Valid values are address and port.
```
Comment (mikewiebe) [inline README.md]: 'address' and 'port'

File: lib/puppet/type/cisco_object_group_entry.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+  newproperty(:address) do
+    desc 'Address to match against. Valid values are an IP'\
+         ' address/prefix_len, IP Address and wildcard, host and '\
+         'host address'
+
+    validate do |address|
+      fail 'address must be ip address/prefix_len,'\
+           ' address/prefix_len, IP Address and wildcard, host and host address' unless
+        %r{host \S+|\S+\/\d+|\S+ [:\.0-9a-fA-F]+}.match(address)
```
Comment (mikewiebe) [inline lib/puppet/type/cisco_object_group_entry.rb]: Might be better to use `PuppetX::Cisco::Utils.process_network_mask()`?

https://github.com/cisco/cisco-network-puppet-module/blob/develop/lib/puppet/type/cisco_bgp_neighbor.rb#L174

**Comment by mikewiebe** [comment]
@saichint Done with first round of review.

File: README.md (hunk 2 above)
Comment (saichint) [inline README.md]: As discussed we need this for destroy

File: README.md (hunk 3 above)
Comment (saichint) [inline README.md]: To keep consistent with cisco_ace etc.

File: lib/puppet/type/cisco_object_group_entry.rb [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  newproperty(:address) do
+    desc 'Address to match against. Valid values are an IP'\
+         ' address/prefix_len, IP Address and wildcard, host and '\
+         'host address'
+
+    validate do |address|
+      addr_arr = address.split
+      addr_arr.each do |addr|
+        next if addr == 'host'
+        PuppetX::Cisco::Utils.process_network_mask(addr)
```
Comment (mikewiebe) [inline lib/puppet/type/cisco_object_group_entry.rb]: Hmm... I know I asked for this change but not sure I like it :).  What is the error message you get in the context of a puppet run when the addr is an invalid ipv4 or ipv6 address?

Comment (saichint) [inline lib/puppet/type/cisco_object_group_entry.rb]: If I give a bad address:
Error: Failed to apply catalog: Parameter address failed on Cisco_object_group_entry[ipv4 address my_addr1 10]: invalid address at /etc/puppetlabs/code/environments/production/manifests/site.pp:8

If I give a bad prefix:
Error: Failed to apply catalog: Parameter address failed on Cisco_object_group_entry[ipv4 address my_addr1 20]: invalid length at /etc/puppetlabs/code/environments/production/manifests/site.pp:13

The error messages are same for ipv6.

**Comment by mikewiebe** [comment]
👍

**Pull request merged**